### ISSUE TITLE
Maintainer & Contributor Guidelines

### ISSUE THREAD

**Issue by jeffbcross** [ReactiveX/RxJS#121]
We should establish some best practices and guidelines for contributors and maintainers of the project. I have some suggestions based on practices that have evolved over the life of the angular project. I'm sure @mattpodwysocki has some ideas based on what has worked with RxJS.

# Prior Art:

 * [AngularJS Contribution Guidelines](https://github.com/angular/angular.js/blob/master/CONTRIBUTING.md) (For all contributors)
 * [Angular 2 Triage and Labels Process](https://github.com/angular/angular/blob/master/TRIAGE_AND_LABELS.md) (For maintainers)
 * [RxJS Contributing Guidelines](https://github.com/Reactive-Extensions/RxJS/tree/master/doc/contributing)

# Common Characteristics
 * Both claim to adhere to Google JavaScript style guide
 * Both emphasize documentation and testing
 * Both provide some guidance on how to design/implement changes

# Philosophies and Values of Angular Team for Consideration. 

These are more focused on work of the maintainers than on community contributors.

 * __Clean, flat commit history__. We never click the green merge button on PRs, but instead we pull down the PR branch and rebase it against master then replace master with the PR branch. See [example gist](https://gist.github.com/jeffbcross/307c6da45d26e29030ef). This reduces noise in the commit history, removing all of the merge commits, and keeps history flat. The flat history is beneficial to tools/scripts that analyze commit ancestry.
 * __Always green master__. Failing master builds tend to cascade into other broken builds, and frustration among other contributors who have rebased against a broken master. Much of our deployment and other infrastructure is based on the assumption that master is always green, nothing should be merged before Travis has confirmed that a PR is green, even for seemingly insignificant changes. Nothing should be merged into a red master, and whomever broke it should drop everything and fix it right away. Fixes should be submitted as a PR and verified as green instead of immediately merging to master.
 * __No force pushes to master__. Only in rare circumstances should a force push to master be made, and other maintainers should be notified beforehand. The most common situation for a justified force push is when a commit has been pushed with an invalid message. The force push should be made as soon as possible to reduce side effects.
 * __Small, logical commits__. A PR should be focused on a single problem, though that problem may be reasonable to be broken into a few logical commits. For example, a global renaming may be best to be broken into a single commit that renames all files, and then a commit that renames symbols within files. This makes the review process simpler easier, so the diff of the meaty commit (where symbols are renamed) can be easily understood than if both were done in the same commit, in which case github would just show a deleted file and an added file.

There's a lot more to our process, but those are more tactical details that can wait until after some consensus is established on core values around contributing.

Additions or alternative perspectives on these values @blesh and @mattpodwysocki?

**Comment by blesh** [comment]
- **No changes to performance tests** You can add new performances tests, but you can't change the old ones without *serious* justification. This is because we want to test outcomes from code changes vs what they were running against. If you *must* change a perf test, it should be done in it's own PR.

**Comment by thejameskyle** [comment]
> - Always green master
> - No force pushes to master

You might want to use GitHub's [protected branches](https://help.github.com/articles/about-protected-branches/) to enforce these.

**Comment by blesh** [comment]
@thejameskyle good tip. I didn't know about that feature. `master` is now protected from force push.

**Comment by blesh** [comment]
Okay, I've got a start in, mostly copy pasta from this issue with a tip of the hat to @jeffbcross embedded for fun.

https://github.com/ReactiveX/RxJS/commit/a3f3746ab1cac095c72a0e6cfa1fab13087ecb01

Closing.
